app/logic/minor.py

```diff
@@ -68,34 +111,171 @@ def getProgramEngagementHistory(program_id, username, term_id):
+    """
+
+    if engagementData['type'] not in ['program','course']:
+        raise Exception("Invalid engagement type!")
+
+    requirement = (CertificationRequirement.select()
+                                           .join(IndividualRequirement, JOIN.LEFT_OUTER, on=((IndividualRequirement.requirement == CertificationRequirement.id) & 
+                                                                                             (IndividualRequirement.username == engagementData['username'])))
+                                           .where(IndividualRequirement.username.is_null(True),
+                                                  CertificationRequirement.certification == Certification.CCE, 
+                                                  CertificationRequirement.name.not_in(['Summer Program'])))
+    if action == 'add':
+        try: 
+            IndividualRequirement.create(**{engagementData['type']: engagementData['id'],
+                                           "username": engagementData['username'],
+                                           "term": engagementData['term'],
+                                           "requirement": requirement.get(),
+                                           "addedBy": currentUser,
+                                        })
+            
+        except DoesNotExist as e:
+            raise e 
+        
+    elif action == 'remove':
+        IndividualRequirement.delete().where(
+                getattr(IndividualRequirement, engagementData['type']) == engagementData['id'],
+                IndividualRequirement.username == engagementData['username'],
+                IndividualRequirement.term == engagementData['term']
+            ).execute()
+    else:
+        raise Exception(f"Invalid action '{action}' sent to setCommunityEngagementForUser")
 
 
 def getCommunityEngagementByTerm(username):
     """
     Given a username, return all of their community engagements (service learning courses and event participations.)
     """
-    courses = (Course.select(Course)
+    courseMatchCase =  Case(None, [(IndividualRequirement.course.is_null(True) , 0)], 1)
+
+    courses = (Course.select(Course, courseMatchCase.alias("reqMatch"))
                      .join(CourseParticipant, on=(Course.id == CourseParticipant.course))
+                     .join(IndividualRequirement, JOIN.LEFT_OUTER, on=((IndividualRequirement.course == Course.id) & 
+                                                                       (IndividualRequirement.username == CourseParticipant.user) & 
+                                                                       (IndividualRequirement.term == Course.term)))
                      .where(CourseParticipant.user == username)
                      .group_by(Course.courseName, Course.term))
     
     # initialize default dict to store term descriptions as keys mapping to each
     # engagement's respective type, name, id, and term.
     terms = defaultdict(list)
```

Comment:
I think terms is a non descriptive variable name. I think something like `communityEngagementByTermDict` or something similar would be better and more descriptive.